mlflow/projects/__init__.py

```diff
@@ -97,20 +100,29 @@ def run(uri, entry_point="main", version=None, parameters=None, experiment_id=No
                   the current process will block when exiting until the local run completes.
                   If the current process is interrupted, any asynchronous runs launched via this
                   method will be terminated.
-    :return: A `SubmittedRun` exposing information (e.g. run ID) about the launched run.
+    :param run_id: Note: this argument is used internally by the MLflow project APIs and should
+                   not be specified. If specified, the given run ID will be used instead of
+                   creating a new run.
+    :return: A `SubmittedRun` exposing information (e.g. run ID) about the launched run. Note that
+             the returned `SubmittedRun` is not thread-safe.
     """
+    if mode == MLFLOW_ENTRY_POINT_MODE:
```

Comment:
`projects.run` calls `_run` which calls `_run_local` which calls `_run_project` which calls `_launch_local_run` which calls `subprocess.Popen` with a special mode which calls `projects.run` which bypasses the `_run` invocation to call `_run_local_noenv`.

That's a pretty long chain to understand, and most of those methods are only used by the method immediately prior to them. This suggests that we could refactor them into a driver method, and a bunch of helper methods to make the driver method more readable.